src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts

```diff
@@ -61,40 +62,118 @@ export class ChatComponent implements OnInit, AfterViewInit, OnDestroy {
     this.isMobileView = event.outerWidth < 530;
   }
 
-  constructor(
-    private store: Store,
-    private route: ActivatedRoute,
-    private location: Location
-  ) {}
+  constructor(private store: Store, private route: ActivatedRoute, private location: Location, private signalRService: SignalRService) {}
 
   ngOnInit(): void {
     this.addNavPath();
-    this.getChatRoom();
     this.getUserRole();
-    this.addListeners();
+    this.getChatRoom();
+    this.createHubConnection();
     this.onResize(window);
   }
 
   ngAfterViewInit(): void {
-    //TODO: Replace setTimeout
-    setTimeout(() => {
-      this.chatEl.nativeElement.scrollTop =
-        this.chatEl.nativeElement.scrollHeight;
-    }, 100);
+    this.makeChatSubscriptions();
+
+    this.chatEl.nativeElement.onscroll = () => {
+      if (this.chatEl.nativeElement.scrollTop < 10) {
+        this.chatEl.nativeElement.scrollTop = 10;
+
+        if (!this.isAdditionalMessageLoading) {
+          this.isAdditionalMessageLoading = true;
+          this.store.dispatch(new GetChatRoomMessages(this.chatRoom.id, this.userRole, { from: this.messages.length, size: 20 }));
+        }
+      }
+    };
   }
 
-  getChatRoom() {
-    const chatRoomId = this.route.snapshot.queryParams['chatRoomId'];
-    this.store.dispatch(new GetChatRoomById(chatRoomId));
+  getChatRoom(): void {
+    const mode = this.route.snapshot.queryParamMap.get('mode');
+
+    switch (mode) {
+      case 'companion':
+        this.createChatRoom();
+        break;
+      default:
+        const chatRoomId = this.route.snapshot.paramMap.get('id');
+        this.store.dispatch(new GetChatRoomById(chatRoomId, this.userRole));
+    }
   }
 
-  getUserRole() {
+  createChatRoom(): void {
+    const companionId = this.route.snapshot.paramMap.get('id');
+    this.chatRoom = {
+      id: null,
+      parentId: null,
+      workshopId: null,
+      workshop: null,
+      parent: null
+    };
+
+    this.store.dispatch(new GetWorkshopById(companionId));
+    this.workshop$
+      .pipe(
+        filter((workshop: Workshop) => !!workshop),
+        takeUntil(this.destroy$)
+      )
+      .subscribe((workshop: Workshop) => {
+        this.chatRoom.workshop = workshop;
```

Comment:
are you sure you need to create a chatRoom before having workshop data? I would like 1) to use a class constructor, 2) to assign this.chatRooom in subscription. Alsom I see on 131 lien you assign parentId as user id, buuut if your role is provider, than it should not happen, correct?

Reply:
and probably this is my fault, but there is a require,ent, that only parent can start a chatRoom (can send first message), because there is no way for provider to find some parent and send a message.